paasta_tools/contrib/timeouts_metrics_prom.py

```diff
@@ -0,0 +1,59 @@
+#!/opt/venvs/paasta-tools/bin/python
+"""
+The goal is to walk through all smartstack.yaml in yelpsoa directories
+and convert timeout values to the prometheus format.
+
+For every upstream and path, if endpoint timeout is specified, we record
+per endpoint timeout value. If not and if timeout_server_ms is specified,
+we record timeout_server_ms value in path `default`. Otherwise, we record the default
+timeout_server_ms 1000.
+
+# HELP endpoint timeout value defined in yelpsoa-config.
+# TYPE TYPE yelpsoaconfig_endpoint_timeouts_ms gauge
+yelpsoaconfig_endpoint_timeouts_ms{path="/consumer_app/devices/braze/info",upstream="push_notifications.main.egress_cluster"} 10000.0
+yelpsoaconfig_endpoint_timeouts_ms{path="default",upstream="mysql_read_security.main.egress_cluster"} 100.0
+"""
+import os
+
+import yaml
+from prometheus_client import CollectorRegistry
+from prometheus_client import Gauge
+from prometheus_client import write_to_textfile
+
+SOA_DIR = "/nail/etc/services/"
+PROM_OUTPUT_FILE = "/nail/etc/services/.autotune_timeouts.prom"
+
+
+def read_and_write_timeouts_metrics(root, service, prom_metric):
+    with open(os.path.join(root, "smartstack.yaml")) as smartstack_file:
+        smartstack_yaml = yaml.safe_load(smartstack_file)
+    for instance_name, info in smartstack_yaml.items():
+        upstream = service + "." + instance_name + ".egress_cluster"
+        if "endpoint_timeout" in info:
+            for path, endpoint_timeout in info["endpoint_timeout"].items():
```

Comment:
hmmm I'm just thinking about how currently a lot of the endpoint timeouts that are set in configs are more generic because regex-matching wasn't available -- and so some of the endpoint timeouts in configs don't necessarily match what we use from pulling the swagger docs. 

like for example in configs if a timeout is set for `/bubbles/` (fake example) that timeout also applies to everything under `/bubbles/` too right? whereas in swagger docs, the list of endpoints we might actually be working with is `/bubbles/v1/blossom/` and `/bubbles/v1/buttercup` or something.

i don't know if _here_ while scraping the metrics is the best place to make sure the lists match though? that might be really hard to do. the easier thing to do might just be to set different dashboard variables on grafana (one set for the swagger endpoints, one set for these config endpoints)

Reply:
I'm not _entirely_ sure whether the regex field in grafana accepts using other variables (I know the query field does, and it allows regex-ing those variables too). I don't have a method for testing this what we have right now, but it's not mentioned in the docs [docs link here](https://grafana.com/docs/grafana/latest/variables/filter-variables-with-regex/)

That being said I also realised we'd have to use another set of dashboard variables either way because the metrics are being scraped to a different shard, so maybe this is a non-issue.

Reply:
I will look into the regex regex-ed variable query using current metrics. Hm I don't know if we need `another set of dashboard variable`, does it mean the current variable (path, upstream) we scrape from  perel won't be able to used in query in compute-infra... I suspect it is the case, I will double check on it...

Reply:
ok I tried and regex definitely works with `path=~".*$regex_path.*|/"` where `regex_path` will be the `yelpsoaconfig_endpoint_timeouts_ms`'s path.
I simulate `regex_path` query variable with regexed`\^(?<path>\S+)/\?\$` of `label_values(upstream_path_status:service_timings_seconds_bucket:rate30s_sum{upstream=~"$upstream"}, path)`
https://grafana.yelpcorp.com/d/Ug6GreD7z/timeouts-playground-copy?orgId=1&var-upstream=ad_delivery_java.main.egress_cluster&var-path=%5E%2Fv1%2Fquery_understanding%2F%3F$&var-cluster=pnw&var-regex_cluster=nova.*&var-regex_cluster_query=pnw&var-regex_path=%2Fv1%2Fquery_understanding

Reply:
Ning and I talked it out, and we figured out we'll have three dashboard variables:
- upstream (currently there)
- regex_path (pulls from these metrics, hidden to the user)
- path (pulls from swagger metrics, will be filtered based on regex_path)

Resolving this conversation now that we've got that sorted :)